qt-bugs@ issue : N211089
Trolltech task ID : 211301
bugs.kde.org number : None
applied: no
author: Dirk Mueller <mueller@kde.org>

when launching Qt3 applications under a KDE that runs with Qt 4.4,
all apps either hang, crash, or eat horrible amounts of memory.

--- src/gui/kernel/qpalette.cpp
+++ src/gui/kernel/qpalette.cpp
@@ -79,9 +79,9 @@ QDataStream &qt_stream_out_qcolorgroup(Q
         int max = QPalette::NColorRoles;
         if (s.version() <= QDataStream::Qt_2_1)
             max = QPalette::HighlightedText + 1;
-        if (s.version() <= QDataStream::Qt_4_3)
+        else if (s.version() <= QDataStream::Qt_4_3)
             max = QPalette::AlternateBase + 1;
-        for(int r = 0 ; r < max ; r++)
+       for(int r = 0 ; r < max ; r++)
             s << g.brush((QPalette::ColorRole)r);
     }
     return s;
@@ -1021,7 +1021,7 @@ QDataStream &operator<<(QDataStream &s,
             int max = QPalette::ToolTipText + 1;
             if (s.version() <= QDataStream::Qt_2_1)
                 max = QPalette::HighlightedText + 1;
-            if (s.version() <= QDataStream::Qt_4_3)
+            else if (s.version() <= QDataStream::Qt_4_3)
                 max = QPalette::AlternateBase + 1;
             for (int r = 0; r < max; r++)
                 s << p.d->br[grp][r];
--- src/gui/painting/qbrush.cpp
+++ src/gui/painting/qbrush.cpp
@@ -924,12 +924,24 @@ QDebug operator<<(QDebug dbg, const QBru
 
 QDataStream &operator<<(QDataStream &s, const QBrush &b)
 {
-    s << (quint8)b.style() << b.color();
-    if (b.style() == Qt::TexturePattern) {
+    quint8 style = (quint8) b.style();
+
+    if (s.version() < QDataStream::Qt_4_0) {
+
+        if (style == Qt::LinearGradientPattern
+                || style == Qt::RadialGradientPattern
+                || style == Qt::ConicalGradientPattern)
+           style = Qt::NoBrush;
+    }
+
+    s << style << b.color();
+    if (b.style() == Qt::TexturePattern)
         s << b.texture();
-    } else if (b.style() == Qt::LinearGradientPattern
+
+    if (s.version() >= QDataStream::Qt_4_0 && (
+            b.style() == Qt::LinearGradientPattern
                || b.style() == Qt::RadialGradientPattern
-               || b.style() == Qt::ConicalGradientPattern) {
+               || b.style() == Qt::ConicalGradientPattern)) {
         const QGradient *gradient = b.gradient();
         int type_as_int = int(gradient->type());
         s << type_as_int;
